### PR TITLE
Removing more Boost dependencies

### PR DESCRIPTION
**Pull request by hkaiser**
- make use of `boost::exception` (and related facilities) conditional
- replace `boost::counting_iterator`
- replace `boost::io::ios_flags_saver`
- flyby: fully remove the use of `boost::exception` where not needed
- flyby: HPX_WITH_DYNAMIC_HPX_MAIN is not supported on windows

Working towards #3440

### CODE REVIEW

File: hpx/runtime/parcelset/locality.hpp [hunk 1]
```diff
@@ -1,6 +1,7 @@
 //  Copyright (c)      2014 Thomas Heller
 //  Copyright (c) 2007-2017 Hartmut Kaiser
 //  Copyright (c) 2007 Richard D. Guidry Jr.
+//  Copyright 2002, 2005 Daryle Walker
```
Comment (K-ballo) [inline hpx/runtime/parcelset/locality.hpp]: This looks out of place.

Comment (hkaiser) [reply hpx/runtime/parcelset/locality.hpp]: Yah, that's not right. Thanks for spotting.

**Comment by hkaiser** [comment]
I believe the errors on gcc-oldest are unrelated to this PR. It is ready to be reviewed now.

File: CMakeLists.txt [hunk 2]
```diff
@@ -288,7 +288,7 @@ endif()
 hpx_option(HPX_WITH_DYNAMIC_HPX_MAIN BOOL
   "Enable dynamic overload of system ``main()`` (Linux only, default: ON)"
   ON ADVANCED)
-if(HPX_WITH_DYNAMIC_HPX_MAIN)
+if(HPX_WITH_DYNAMIC_HPX_MAIN AND (NOT MSVC))
```
Comment (msimberg) [inline CMakeLists.txt]: Should we also make it a hard configuration error if `HPX_WITH_DYNAMIC_HPX_MAIN=ON` on Windows? There might be quite a few others as well in which case we can do this separately?

Comment (hkaiser) [reply CMakeLists.txt]: Let's do this separately. I agree we should turn non-applicable options into hard configuration errors.

**Pull request merged**